### DISCUSSION TITLE
Is there a way to detect when the visible window is closed by the user?

### DISCUSSION THREAD

**Discussion by postmodern** [rubycdp/ferrum#383] [Q&A] [answered]
Hello, I want to open a visible instance of `Ferrum::Browser`, leave it open for the user, and wait for the user to then close the browser window, then call `browser.quit` and exit from Ruby. Is there some kind of `wait_for_close` callback event?

**Comment by postmodern** [comment]
After initializing `Ferrum::Browser` with a Debug logger that printed every debug message, it appears that a `Inspector.detached` message is sent back when the Chrome window is closed.

```ruby
browser.on('Inspector.detached') do
  ...
end
```

I still don't have a good way to wait until the `Inspector.detached` message is sent, and let other `on()` callbacks to be called.

**Comment by postmodern** [comment, marked as answer]
Since every `browser.on(...) { ... }` block is executed using Concurrent Ruby, one cannot simply call `exit` as all exceptions are silently rescued within the blocks; and `exit` raises `SystemExit`. However, you can use this really simple solution that sets a local variable and busy waits until it's set:

```ruby
window_closed = false

browser.on('Inspector.detached') do
  window_closed = true
end

until window_closed
  sleep(1)
end

browser.quit
exit
```

**Reply by route** [reply]
First of all I got rid of Async and now we raise errors in the threads. Second I think `Target.targetDestroyed` is closer to the Page/Tab/Window being closed.
